From 340b2a071c98deef29a96ec9275bc84092b742d8 Mon Sep 17 00:00:00 2001
From: Dmitry <docent27@ukr.net>
Date: Sun, 11 Jul 2021 13:47:46 +0300
Subject: [PATCH] Shared Libary

---
 Android.mk | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/Android.mk b/Android.mk
index bb97245..53daf6e 100644
--- a/Android.mk
+++ b/Android.mk
@@ -95,8 +95,7 @@ LOCAL_CFLAGS += -DNO_GBM
 #LOCAL_CFLAGS += -DNO_INIT_CONSTRUCTOR
 LOCAL_CFLAGS += -DDEFAULT_ES=2
 
+LOCAL_CFLAGS += -Dasm=__asm__ -Dvolatile=__volatile__
 LOCAL_LDLIBS := -llog
-#building as a static lib
 
-LOCAL_CFLAGS += -DSTATICLIB
-include $(BUILD_STATIC_LIBRARY)
+include $(BUILD_SHARED_LIBRARY)
--
libgit2 0.28.4

