From 7d85d75c97bcc106935ecc11842419950f8b91ca Mon Sep 17 00:00:00 2001
From: Docent27 <docent27@ukr.net>
Date: Sun, 26 Sep 2021 11:34:07 +0300
Subject: [PATCH] mygui-selector

---
 apps/openmw/engine.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/apps/openmw/engine.cpp b/apps/openmw/engine.cpp
index a06faa4..ae460a2 100644
--- a/apps/openmw/engine.cpp
+++ b/apps/openmw/engine.cpp
@@ -704,7 +704,11 @@ void OMW::Engine::prepareEngine (Settings::Manager & settings)
     else
         gameControllerdb = ""; //if it doesn't exist, pass in an empty string
 
-    std::string myguiResources = (mResDir / "mygui").string();
+    std::string myguiDir = mResDir.string() + "/mygui";
+    const char *path = getenv("OPENMW_MYGUI");
+    if(path) 
+    myguiDir = myguiDir + "/" + path;
+    std::string myguiResources = (myguiDir);
     osg::ref_ptr<osg::Group> guiRoot = new osg::Group;
     guiRoot->setName("GUI Root");
     guiRoot->setNodeMask(MWRender::Mask_GUI);
--
libgit2 0.28.4

