From 7a1c0c2f449f31237ab09e91f2c16fbf83076418 Mon Sep 17 00:00:00 2001
From: Docent27 <docent27@ukr.net>
Date: Mon, 27 Sep 2021 12:52:24 +0300
Subject: [PATCH] water

---
 apps/openmw/mwrender/water.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/apps/openmw/mwrender/water.cpp b/apps/openmw/mwrender/water.cpp
index c5fd1a3..9b9c909 100644
--- a/apps/openmw/mwrender/water.cpp
+++ b/apps/openmw/mwrender/water.cpp
@@ -627,8 +627,8 @@ void Water::createShaderWaterStateSet(osg::Node* node, Reflection* reflection, R
     defineMap.insert(std::make_pair(std::string("refraction_enabled"), std::string(refraction ? "1" : "0")));
 
     Shader::ShaderManager& shaderMgr = mResourceSystem->getSceneManager()->getShaderManager();
-    osg::ref_ptr<osg::Shader> vertexShader (shaderMgr.getShader("water_vertex.glsl", defineMap, osg::Shader::VERTEX));
-    osg::ref_ptr<osg::Shader> fragmentShader (shaderMgr.getShader("water_fragment.glsl", defineMap, osg::Shader::FRAGMENT));
+    osg::ref_ptr<osg::Shader> vertexShader (shaderMgr.getShader(getenv("OPENMW_WATER_VERTEX"), defineMap, osg::Shader::VERTEX));
+    osg::ref_ptr<osg::Shader> fragmentShader (shaderMgr.getShader(getenv("OPENMW_WATER_FRAGMENT"), defineMap, osg::Shader::FRAGMENT));
 
     osg::ref_ptr<osg::Texture2D> normalMap (new osg::Texture2D(readPngImage(mResourcePath + "/shaders/water_nm.png")));
 
--
libgit2 0.28.4

